packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx

```diff
@@ -0,0 +1,192 @@
+
+import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
+import { getEncoding } from 'js-tiktoken';
+import yaml from 'js-yaml';
+import _ from 'lodash';
+import { UseToolsProps } from '../../use-tools.js';
+import { cohereContext, reranker } from '../../../lib/cohere.js';
+
+export type LengthFunction = ((text: string) => number) | ((text: string) => Promise<number>);
+export interface RedactedFuncionResponseMetadata {
+  isRedacted: true;
+  chunks: string[];
+}
+
+const getEncoder = _.once(() => getEncoding('cl100k_base'));
+
+function openAITokenCount(text: string) {
+  return getEncoder().encode(text).length;
+}
+
+export async function TruncateByChars(
+  {
+    children,
+    maxLength,
+  }: {
+    children: string;
+    maxLength: number;
+  },
+  { render }: AI.ComponentContext
+) {
+  const stringified = await render(children);
+  if (stringified.length <= maxLength) {
+    return stringified;
+  }
+  return `${stringified.slice(0, maxLength - 3)}...`;
+}
+
+export async function LargeFunctionResponseHandler(
+  {
+    children,
+    maxLength,
+    numChunks = 4,
+    // use Cohere token counter?
+    lengthFunction = openAITokenCount,
+    ...props
+  }: AI.PropsOfComponent<typeof FunctionResponse> & {
+    maxLength: number;
+    numChunks?: number;
+    lengthFunction?: LengthFunction;
+  },
+  { render, logger, getContext }: AI.ComponentContext
+) {
+  let stringified = await render(children);
+
+  // Option 1: do nothing if it's already small enough
+  if ((await lengthFunction(stringified)) <= maxLength) {
+    return <FunctionResponse {...props}>{stringified}</FunctionResponse>;
+  }
+
+  stringified = yamlOptimizeIfPossible(stringified);
+
+  // Option 2: try dumping as YAML, if it's small enough then we are done
+  if ((await lengthFunction(stringified)) <= maxLength) {
+    return <FunctionResponse {...props}>{stringified}</FunctionResponse>;
+  }
+
+  // Option 3 (last reosrt): split into chunks and allow LLM to query by similarity
+  // Requires Cohere API key for doing similarity search
+  const cohereConfig = getContext(cohereContext);
+
+  if (!cohereConfig.api_key) {
+    // Not yet an error, but this is an error just waiting to happen
+    logger.warn(
+      { CohereContext: cohereConfig },
+      'FunctionResponse is too big, but Cohere API key is not set. Please set it in the context.'
+    );
+  }
+
+  const splitter = new RecursiveCharacterTextSplitter({
+    chunkSize: maxLength / numChunks,
+    chunkOverlap: maxLength / numChunks / 10,
+    lengthFunction,
+  });
+  const chunks = await splitter.splitText(stringified);
+
+  return (
+    <FunctionResponse {...props} metadata={{ isRedacted: true, chunks }}>
+      ... The response is too big and hence redacted. The response can be queried using semantic similarity search by
+      calling the `loadBySimilarity` function.
+    </FunctionResponse>
+  );
+  // TODO: return a chunk based on query (gotta get it from conversation)
```

Comment:
I'd rather not to do this, but @juberti if you still want it I can take another stab at it.